### DISCUSSION TITLE
Connector: is there any style or option for the end of the connector to point to start of the node?

### DISCUSSION THREAD

**Discussion by biswajit-goswami** [freeplane/freeplane#1136] [Help] [answered]
When I connect two nodes , the end of the arrow points to the end of the destination node. I can drag the the connector to make it point to the start of the node..but is there some option which can always make the arrow to connect to the beginning of the destination node?

**Comment by dpolivaev** [comment]
Edit connector properties for the style "default" connector in the style editor.
For new maps, edit your default template map.

**Reply by biswajit-goswami** [reply]
where can I edit the connector properties to achieve this behaviour?
I tried the tools-> preferences...as well as the edit styles....did not see any option to do this
![image](https://github.com/freeplane/freeplane/assets/11291623/1a95d54d-db57-459a-91d8-8956ffcd26ae)

**Reply by EdoFro** [reply]
I think you can do this only with scripting. 
Look at the API (ctrl + F1: search for menu command “api”). 
You get the object connector through the node, and the connector has a method to set its inclination points. Setting one of them with negative values makes the connector to enter the node by the other side. 
I’m not at my computer now to code an example script. Maybe later. 
hope this helps
edo

**Reply by biswajit-goswami** [reply]
@EdoFro , if you can, plz share the script....also, could this approach be replicated for the whole map...ie any connector in the map shows this behavior?

**Reply by dpolivaev** [reply]
Actually, the drop-down box "Connector arrows" shown on the screenshot should offer all possible options.

**Comment by macmarrum** [comment, marked as answer]
The way I read the question "Connector: is there any style or option for the end of the connector to point to start of the node?", it is about attachment points.

Here's an example

![connector-from-end-to-beginning xi](https://github.com/freeplane/freeplane/assets/7197131/84eefa61-b864-4071-b82f-950584cb884b)

To achieve `(A end) to (beginning C)`, I created a connector between **A** and **C**, with **Connector arrows:** `Draw arrow forward`, then I ran the following script on **C** to change the inclination points, as suggested by Edo: `node.connectorsIn.each { it.setInclination([100, 0], [-100, 0]) }`

It would be great to have a relevant option in the **Connector dialog**, e.g.

**Attachment points:** 
- Node end - node end
- Node end - node beginning
- Node beginning - node end
- Node beginning - node beginning

[connector-from-end-to-beginning.xi.mm.zip](https://github.com/freeplane/freeplane/files/11468895/connector-from-end-to-beginning.xi.mm.zip)

**Reply by hyouriku** [reply]
Thanks a lot! I was looking after this for months now (not finding it till now).
I happen to have a small cosmetic issue when using the script. Is there a way to define the shape of the connector? For example if you use a connector between (A end) and (B end) to have it create the shape of a semicircle instead of choosing a more direct/ellyptic connection?

**Comment by biswajit-goswami** [comment]
this approach works...but in order for this to be applicable on all nodes in the map, should the attribute be added to all the nodes? and is there a quick way to do it?
I would prefer this as an option in the connector settings, as @macmarrum mentioned above.

**Reply by macmarrum** [reply]
You can run the script for all nodes in your mindmap. 

One way to do it is to create the following script on any node and run it:
```groovy
// @ExecutionModes({ON_SINGLE_NODE})
node.mindMap.root.findAll().each { n -> 
    n.connectorsIn.each { it.setInclination([100, 0], [-100, 0]) }
}
```

Another way is to save the script as a file in `Tools->Open user directory` -> `scripts`, e.g. named `setAllConnectorsEndToBeginning.groovy`, and then run it from `Tools->Scripts->Set All Connectors End To Beginning`

More info about scripting in general: https://docs.freeplane.org/scripting/Scripting.html

**Reply by biswajit-goswami** [reply]
Yes, it works .
Also, if I want to make the arrow start from the mid of the node to connect to the mid of the destination node, what changes will be required?
I checked the API doc, but was not able to derive the right values to do so:
![image](https://github.com/freeplane/freeplane/assets/11291623/312e5f73-3b5e-4341-a99b-4b75628f63ca)

**Comment by euu2021** [comment]
For future reference:

This script creates a Connector and immediately sets the inclination and anchor points, as desired:
`node.addConnectorTo(c.selecteds[1]).setInclination([90,00], [-180,00])`

I set it to be: connectors always leave from the right side of the node and income on the left side of the node.
It respects the order that the nodes were selected to decide which will be the source node, and which wil be the target node.

For comparison:
connecting aaa and bbb using the script: 

![image](https://github.com/freeplane/freeplane/assets/77707706/727d3c0e-3aba-4f23-ab56-9f16713adb99)

While using the default `Insert->Connect`, it would be like:

![image](https://github.com/freeplane/freeplane/assets/77707706/14f6c81e-d97c-4a04-8a7b-05d569e1875b)

**Reply by euu2021** [reply]
Actually, now I see that it's possible to choose the default inclination and anchor points in the Styles (as mentioned in the first comment https://github.com/freeplane/freeplane/discussions/1136#discussioncomment-5884394). No scripts needed. And FP is even smart enough to make the proper inversion when the nodes are on the left side 👏

So, just set it in the Styles, like this:
![image](https://github.com/freeplane/freeplane/assets/77707706/f42b389a-aa51-41f8-95aa-5339e83eedee)

**Reply by biswajit-goswami** [reply]
@euu2021 , I don't see such option in FP 1.11.8 version. In which version the option is available?

**Reply by euu2021** [reply]
Go into the Styles manager (`Format->Manage Styles->Edit styles`). You will see a window like this:
![image](https://github.com/freeplane/freeplane/assets/77707706/f69b2114-90c3-49e9-98cd-92c062851756)

There, you can change the connector, to make it how you like it. For example, to make it default to income at the top, and outgo at the bottom:
![image](https://github.com/freeplane/freeplane/assets/77707706/3ea68975-841b-4617-818d-052fd39fd685)

**Reply by biswajit-goswami** [reply]
so, we need to drag the orientation of the connector in the styles. I thought there was some drop down, etc. Thanks.

**Reply by Lotus49b** [reply]
Is there a specific action to take when dragging the start and end points around the "default" node, such as hold down Ctrl key or similar)?
In my setup (ver 1.11.12-pre04) the connector start point swings around to any chosen side of the node (e.g. bottom of default node), but I can't get the end-point to attach to a chosen side (e.g. top of default node). I've tried dragging the end point, holding down ctrl keys, etc.

**Reply by euu2021** [reply]
I think the trick is making small movements:

![javaw_Zo2nPOlKYt](https://github.com/freeplane/freeplane/assets/77707706/6d87e7a5-46be-407f-8f8e-5a31f2dfb0c5)


I think it can be called fly fishing technique 🤣

**Reply by Lotus49b** [reply]
Thank you for you kind reply.
For some reason my default node connector is a simple, single end-point line, and not looped back to the node as yours is.
I'll have to try find the reason for that. 
![Screenshot 2024-03-19 204134](https://github.com/freeplane/freeplane/assets/11365744/0738e8a9-3868-4e72-8dfc-4fbb1a4caded)
In the meantime, I have noticed that the connector can be relocated in the live mind map, so that's already a win!

**Reply by euu2021** [reply]
Oh, I see. You need to:
- right-click the connector and set it to `Connector Shape -> Curve`
- set it the directions you want
- change it back to `Connector Shape -> Line`. It will behave the way you set in the step above.

Connectors are tricky 🤣
